pkg/cmd/roachtest/tests/disk_stall.go

```diff
@@ -15,125 +15,441 @@ import (
+			Cluster: r.MakeClusterSpec(4),
+			Timeout: 15 * time.Minute,
+			Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+				runDiskStalledDetection(ctx, t, c, makeStaller(t, c))
+			},
+			// Encryption is implemented within the virtual filesystem layer,
+			// just like disk-health monitoring. It's important to exercise
+			// encryption-at-rest to ensure there is not unmonitored I/O within
+			// the encryption-at-rest implementation that could indefinitely
+			// stall the process during a disk stall.
+			EncryptionSupport: registry.EncryptionMetamorphic,
+		})
+	}
+
+	for _, stallLogDir := range []bool{false, true} {
+		for _, stallDataDir := range []bool{false, true} {
 			// Grab copies of the args because we'll pass them into a closure.
 			// Everyone's favorite bug to write in Go.
-			affectsLogDir := affectsLogDir
-			affectsDataDir := affectsDataDir
+			stallLogDir := stallLogDir
+			stallDataDir := stallDataDir
 			r.Add(registry.TestSpec{
+				Skip:        "#95886",
+				SkipDetails: "The test current only induces a 50us disk-stall, which cannot be reliably detected.",
 				Name: fmt.Sprintf(
-					"disk-stalled/log=%t,data=%t",
-					affectsLogDir, affectsDataDir,
+					"disk-stalled/fuse/log=%t,data=%t",
+					stallLogDir, stallDataDir,
 				),
 				Owner:   registry.OwnerStorage,
-				Cluster: r.MakeClusterSpec(1),
+				Cluster: r.MakeClusterSpec(4),
 				Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
-					runDiskStalledDetection(ctx, t, c, affectsLogDir, affectsDataDir)
+					runDiskStalledDetection(ctx, t, c, &fuseDiskStaller{
+						t:         t,
+						c:         c,
+						stallLogs: stallLogDir,
+						stallData: stallDataDir,
+					})
 				},
+				EncryptionSupport: registry.EncryptionMetamorphic,
 			})
 		}
 	}
 }
 
-func runDiskStalledDetection(
-	ctx context.Context, t test.Test, c cluster.Cluster, affectsLogDir bool, affectsDataDir bool,
-) {
-	if c.IsLocal() && runtime.GOOS != "linux" {
-		t.Fatalf("must run on linux os, found %s", runtime.GOOS)
-	}
+func runDiskStalledDetection(ctx context.Context, t test.Test, c cluster.Cluster, s diskStaller) {
+	const maxSyncDur = 10 * time.Second
+	startOpts := option.DefaultStartOpts()
+	startSettings := install.MakeClusterSettings()
+	startSettings.Env = append(startSettings.Env,
+		"COCKROACH_AUTO_BALLAST=false",
+		fmt.Sprintf("COCKROACH_LOG_MAX_SYNC_DURATION=%s", maxSyncDur),
+		fmt.Sprintf("COCKROACH_ENGINE_MAX_SYNC_DURATION_DEFAULT=%s", maxSyncDur))
 
-	n := c.Node(1)
+	t.Status("setting up disk staller")
+	s.Setup(ctx)
+	defer s.Cleanup(ctx)
+	c.Run(ctx, c.All(), "mkdir -p logs-external")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/cockroach.stdout.log logs/cockroach.stdout.log || true")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/cockroach.stderr.log logs/cockroach.stderr.log || true")
+	c.Run(ctx, c.All(), "ln -s /home/ubuntu/logs-external/roachprod.log logs/roachprod.log || true")
```

Comment:
Is symlinking specific files thorough enough? We may want to place all of the log files on a separate device that can be stalled entirely, to ensure that all log files and log rotation logic and such is also tested properly.

Reply:
This is actually exempting these log files from the stalling. I'll remove this; it was a change made while we suspected that the reason the process wouldn't terminate was that `cockroach.sh` redirected stdout and stderr to the stalled disk. This turned out to not be the reason, and we've realized there's no way to reliably terminate the process.